src/adapters/philips_hue/hub.rs

```diff
@@ -0,0 +1,111 @@
+        info!("Starting Hue Hub Service for {}", self.id);
+        let adapter = self.adapter.clone();
+        let id = self.id.clone();
+        let api = self.api.clone();
+
+        thread::spawn(move || {
+
+            // The main Hub management loop
+            loop {
+                if !api.is_available() {
+                    // Re-check availability every minute.
+                    thread::sleep(Duration::from_millis(60*1000));
+                    continue;
+                }
+
+                // If the Hub is not paired, try pairing.
+                if !api.is_paired() {
+                    debug!("Push pairing button on Philips Hue Bridge ID {}", id);
+
+                    // Try pairing for 120 seconds.
+                    for _ in 0..120 {
+                        adapter.controller.adapter_notification(
+                            json_value!({ adapter: "philips_hue",
+                                message: "NeedsPairing", hub: id }));
+                        if api.try_pairing() {
+                            break;
+                        }
+                        thread::sleep(Duration::from_millis(1000));
+                    }
+
+                    if api.is_paired() {
+                        info!("Paired with Philips Hue Bridge ID {}", id);
+                        adapter.controller.adapter_notification(
+                            json_value!({ adapter: "philips_hue", message: "PairingSuccess",
+                                hub: id }));
+                    } else {
+                        warn!("Pairing timeout with Philips Hue Bridge ID {}", id);
+                        adapter.controller.adapter_notification(
+                            json_value!({ adapter: "philips_hue", message: "PairingTimeout",
+                                hub: id }));
+                        // Giving up for this Hub.
+                        // Re-try pairing every hour.
+                        thread::sleep(Duration::from_millis(60*60*1000));
+                        continue;
+                    }
+                }
+
+                // We have a paired Hub, instantiate the lights services.
+                // Extract and log some info
+                let setting = api.get_settings();
+                let hs = structs::Settings::new(&setting).unwrap(); // TODO: no unwrap
+                info!(
+                    "Connected to Philips Hue bridge model {}, ID {}, software version {}, IP address {}",
+                    hs.config.modelid, hs.config.bridgeid, hs.config.swversion,
+                    hs.config.ipaddress);
+
+                let light_ids = api.get_lights();
+                for light_id in light_ids {
+                    debug!("Found light {} on hub {}", light_id, id);
+                    adapter.send(HueAction::AddLight(id.to_owned(), light_id.to_owned()));
+                }
+
+                loop { // Forever
+                    thread::sleep(Duration::from_millis(60*1000));
```

Comment:
That looks pretty strange. What's the goal of this infinite loop that does nothing?